### PR TITLE
Introduce deployment scripts for GCP

### PR DESCRIPTION
**Pull request by kadircet**

### CODE REVIEW

**Review by kirillbobyrev — COMMENTED** [review]
Bash code looks normal to me, mostly nits, docs & minor comments.

File: README.md [hunk 1]
```diff
@@ -1,15 +1,3 @@
 # llvm-remote-index
 
 [![Index LLVM project](https://github.com/clangd/llvm-remote-index/workflows/Index%20LLVM%20project/badge.svg)](https://github.com/clangd/llvm-remote-index/actions)
-
-# Creating a docker image
```
[the remark this answers is not in the document]
Comment (kadircet) [reply README.md:5]: added descriptions for deployment/ and docker/ folders.

File: deployment/README.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
+# GCP server management scripts
+
+This directory contains scripts used for managing the GCP project. They make use
+of gcloud SDK so you need to install the SDK first, you can find instructions in
```
[the remark this answers is not in the document]
Comment (kadircet) [reply deployment/README.md]: done

File: deployment/README.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+[here](https://cloud.google.com/sdk/docs/install).
+
+## Configuration
+
+### Common
```
[the remark this answers is not in the document]
Comment (kadircet) [reply deployment/README.md:9]: done

File: deployment/README.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+GCP project name, machine type to use for VM instances and common base names can
+be configured through [args.sh](args.sh).
+
+### Docker options
```
[the remark this answers is not in the document]
Comment (kadircet) [reply deployment/README.md]: moving into args.sh instead.

File: deployment/README.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+Environment variables for the docker container can be configured in
+[push_new_docker_image.sh](push_new_docker_image.sh). These contain:
+- How to fetch the index artifact.
+- Indexer project root.
+
+### Serving infra
+
+This is setup by running [initial_setup.sh](initial_setup.sh) script ones. By
```
[the remark this answers is not in the document]
Comment (kadircet) [reply deployment/README.md]: i am not sure what i was trying to say in first sentence too :D feels like i had an interrupt while writing that one and forgot to read again...

as for the environment i meant "deployment environments", i.e. staging and live not sure if there's a better term for it.

File: deployment/README.md [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+with a single VM in each of them. It has a global TCP loadbalancer in front.
+Loadbalancer accepts traffic on port 5900.
+
+Both environments use a TCP healthcheck on port 50051 and they only allow
+ingress to that port.
+
+## Rolling images back/forward
+
+These can be done via [rollout_new_release.sh](rollout_new_release.sh) and
+[rollback_to_release.sh](rollback_to_release.sh) scripts.
```
[the remark this answers is not in the document]
Comment (kadircet) [reply deployment/README.md]: done

File: deployment/README.md [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+### Rolling out new images
+
+`bash rollout_new_release.sh staging` will create a new docker image, pulling
+the binaries from clangd/clangd/releases page, and push it to staging.
```
[the remark this answers is not in the document]
Comment (kadircet) [reply deployment/README.md]: done

File: deployment/README.md [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+`bash rollout_new_release.sh live` will push the latest available docker image
+in GCR into live, e.g. can be used to promote the latest staging image.
```
[the remark this answers is not in the document]
Comment (kadircet) [reply deployment/README.md]: done

File: deployment/README.md [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+### Rolling back to older images
+
+`bash rollback_to_release.sh [staging|live] IMAGE_FQN` can be used to change
+images for staging|live instaces.
+
+Fully qualified image names can be acquired either through GCP web UI or through
```
[the remark this answers is not in the document]
Comment (kadircet) [reply deployment/README.md]: done

File: deployment/args.sh [hunk 10]
```diff
@@ -0,0 +1,28 @@
+#!/bin/bash
+# Abort script on failure and print commands as we execute them.
+set -x -e
+
+# GCP project to configure.
+PROJECT_ID="llvm-remote-index"
+
+# Basename for instance templates, can be suffixed with image shas.
```
[the remark this answers is not in the document]
Comment (kadircet) [reply deployment/args.sh]: done

File: deployment/args.sh [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+BASE_TEMPLATE_NAME="${PROJECT_ID}-server-template"
+
+# Machine type to use for index serving VM instances.
+# 2 vCPUs and 4GB ram is enough for serving llvm-index.
```
[the remark this answers is not in the document]
Comment (kadircet) [reply deployment/args.sh:12]: done

File: deployment/args.sh [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+MACHINE_TYPE="e2-medium"
+
+# Fully qualified name for the server image in google cloud repo.
```
[the remark this answers is not in the document]
Comment (kadircet) [reply deployment/args.sh]: done

File: deployment/args.sh [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+IMAGE_IN_GCR="gcr.io/${PROJECT_ID}/${PROJECT_ID}-server"
+
+# Used as base name for instance groups and machine instances.
+BASE_INSTANCE_NAME="${PROJECT_ID}-server"
+
+# Which github repository to use for fetching index artifacts.
+INDEX_REPO="clangd/llvm-remote-index"
+
+# Prefix to detect index asset.
```
[the remark this answers is not in the document]
Comment (kadircet) [reply deployment/args.sh]: done

File: deployment/args.sh [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+INDEX_ASSET_PREFIX="llvm-index"
+
+# Absolute path to project root on indexer machine.
```
[the remark this answers is not in the document]
Comment (kadircet) [reply deployment/args.sh]: i thought "indexer machine" was enough to point that.

File: deployment/create_vm_template.sh [hunk 15]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+
+if [[ $# -lt 2 ]]; then
+  echo "Usage: $0 FULL_IMAGE TEMPLATE_NAME"
+  echo "  FULL_IMAGE - Full name of the image in GCR."
+  echo "  TEMPLATE_NAME - Name to use for the template."
```
[the remark this answers is not in the document]
Comment (kadircet) [reply deployment/create_vm_template.sh]: done

File: deployment/initial_setup.sh [hunk 16]
```diff
@@ -0,0 +1,45 @@
+#!/bin/bash
+source args.sh
+source setup_utils.sh
+
+# Create one server image and push it to the GCR first and create an instance
```
[the remark this answers is not in the document]
Comment (kadircet) [reply deployment/initial_setup.sh]: done

File: deployment/initial_setup.sh [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+
+# We need two healthchecks, one global for live instance and one regional for
+# staging.
+HEALTH_CHECK_NAME="${BASE_INSTANCE_NAME}-hc-grpc"
+createHC "global" $HEALTH_CHECK_NAME
+createHC "europe-west1" $HEALTH_CHECK_NAME
+
+# Named port for load balancer to redirect traffic into instance groups.
+NAMED_PORT="grpc"
+# Create one staging instance and 2 production instances, with appropriate load
```
[the remark this answers is not in the document]
Comment (kadircet) [reply deployment/initial_setup.sh:26]: done

File: deployment/setup_utils.sh [hunk 18]
```diff
@@ -0,0 +1,98 @@
+#!/bin/bash
+source args.sh
+
+# Creates a TCP health check on port 50051 in specified region with given name.
+function createHC() {
```
[the remark this answers is not in the document]
Comment (kadircet) [reply deployment/setup_utils.sh]: done

File: deployment/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (kirillbobyrev) [reply deployment/README.md:9]: I think you can just remove the `### Common` header keeping the text simply under `## Configuration`.

**Review by kirillbobyrev — COMMENTED** [review]
LGTM with a tiny nit. Thanks for implementing this!

**Review by kirillbobyrev — APPROVED** [review]

**Pull request merged**